pyccel/codegen/printing/fcode.py

```diff
@@ -2000,11 +2000,13 @@ def _print_Is(self, expr):
         if isinstance(expr.rhs, Nil):
             return '.not. present({})'.format(lhs)
 
-        if a.dtype is NativeBool() and b.dtype is NativeBool():
+        if (a.dtype is NativeBool() and b.dtype is NativeBool()):
             return '{} .eqv. {}'.format(lhs, rhs)
 
-        errors.report(PYCCEL_RESTRICTION_IS_RHS, symbol=expr,
-            severity='fatal')
+        if (a.dtype is NativeInteger() and b.dtype is NativeInteger()):
+            return '{} .eq. {}'.format(lhs, rhs)
+
+        raise NotImplementedError("Only booleans, integers and None are currently supported for comparison")
```

Comment:
Why not floats? And as I mentioned before, why not returning `False` whenever `a is b` has different types for `a` and `b`?

Reply:
We didn't think it was a good idea to add floats as this would be quite difficult to achieve if we want the same results as python. The `Is` operator is intrinsically linked to how python stores values in memory. For floats they are usually but not always stored separately. So
```
> a = 3.5
> b = 3.5
> a is b
False
> b = a
> a is b
True
```

Reply:
If the user wants a comparison operator then they should use `==` for floats. `Is` is something different which we can't implement entirely

Reply:
we have the same problem with integers
> a = 275 
> b = 275          
>a is b           
 False

Reply:
I think we should do as numba does by comparing between the two values even if it's does not give us the same result as in python

Reply:
Hmm, it seems you are right. On my computer `is` returns the same thing as `==` for integers between -5 and 256

Reply:
https://stackoverflow.com/questions/15171695/whats-with-the-integer-cache-maintained-by-the-interpreter
you can find here an explanation of the caching system

Reply:
In that case integers and floats should be handled in the same way. I.e, either:
1. By implementing `is` as `==` despite that not being correct behaviour
2. Raise an error saying that pyccel does not handle python's `id` function and therefore cannot implement `is` unless one of the types is either `bool` or `None`
Personally I prefer the second option. What do you think?

Reply:
>https://stackoverflow.com/questions/15171695/whats-with-the-integer-cache-maintained-by-the-interpreter
you can find here an explanation of the caching system

Interesting so -5 and 256 is not just anecdotal then. So technically we could also implement `is` for `np.uint_8` (although we would first have to implement unsigned ints ^^')

Reply:
the second option seems okay for me

Reply:
I also agree, we cannot fully replicate the behavior of `is` in Python, which also changes between interpreters! For example PyPy behaves differently than CPython: https://doc.pypy.org/en/latest/cpython_differences.html#object-identity-of-primitive-values-is-and-id

Reply:
I think that the rules in PyPy ([here](https://doc.pypy.org/en/latest/cpython_differences.html#object-identity-of-primitive-values-is-and-id)) make a lot of sense: for all primitive types `is` is equivalent to `==`, and for other types (arrays for example) it checks object identity (which we can do a priori given that we compile the code statically). What do you guys think?

Reply:
seems good to me

Reply:
do i need to make `is` equivalent to `==` for the the implemented primitive immutable types in this pull request ?

Reply:
yes @nhamidn

Reply:
Personally I am not a fan of this method. All our tests work in the same way: they generate a fortran file and they check that the output of the fortran file and the original python code is the same. If we translate `is` to `==` we will not be able to use this method. To me this feels like an indication that we are doing something contrary to the way pyccel works.

If we do decide to use this method then we will need to either write tests and pass the expected result explicitly, or write 2 functions 1 with `is` to be translated and 1 with `==` for the comparison. The latter method allows us to run automated tests with random values that are more likely to catch corner cases.

If this method were implementing something new then I don't think it would bother me, but as the user has the option to use the `==` operator then if they have chosen to use `is` there should be a reason for that (even if that reason is simply because they are unaware of the difference). Personally I would prefer that we raise an error "The function id() is not implemented for basic types `int`, `float`. `str`. Please use `==` for comparisons"

That said if everyone else thinks it is better to implement `is` following PyPy's rules then I'm happy to go with the majority

Reply:
@EmilyBourne I think you're right here.

I would like to follow PyPy behavior (which is also consistent with Numba's implementation as far as I understand), but this makes little sense as we are using CPython to create our wrappers.

And given that CPython does not handle `id(a)` in a meaningful way when `a` is an immutable primitive type, the only sensible option that remains is exactly what you suggest.

Hence we should raise the error
```python
'Cannot translate "is" comparison, because function "id()" is not implemented for basic types "int", "float", and "str". Please use "==" for comparison instead.'
```

Reply:
```
'Cannot translate "is" comparison, because function "id()" is not implemented for basic types "int", "float", "complex" and "str". Please use "==" for comparison instead.'
```
I missed complex